pkg/reconciler/taskrun/validate_resources.go

```diff
@@ -311,19 +311,27 @@ func validateTaskRunResults(tr *v1beta1.TaskRun, resolvedTaskSpec *v1beta1.TaskS
 }
 
 // mismatchedTypesResults checks and returns all the mismatched types of emitted results against specified results.
-func mismatchedTypesResults(tr *v1beta1.TaskRun, specResults []v1beta1.TaskResult) map[string][]string {
-	neededTypes := make(map[string][]string)
-	providedTypes := make(map[string][]string)
-	// collect needed keys for object results
+func mismatchedTypesResults(tr *v1beta1.TaskRun, specResults []v1beta1.TaskResult) map[string]string {
+	neededTypes := make(map[string]string)
+	mismatchedTypes := make(map[string]string)
+	var filteredResults []v1beta1.TaskRunResult
+	// collect needed types for results
 	for _, r := range specResults {
-		neededTypes[r.Name] = append(neededTypes[r.Name], string(r.Type))
+		neededTypes[r.Name] = string(r.Type)
 	}
 
-	// collect provided keys for object results
+	// collect mismatched types for results, and correct results in filteredResults
 	for _, trr := range tr.Status.TaskRunResults {
-		providedTypes[trr.Name] = append(providedTypes[trr.Name], string(trr.Type))
+		needed, ok := neededTypes[trr.Name]
+		if ok && needed != string(trr.Type) {
+			mismatchedTypes[trr.Name] = fmt.Sprintf("want type: %v, got type: %v", needed, trr.Type)
```

Comment:
suggest a little more verbose and explicit error message:

```
fmt.Sprintf("task result is expected to be \"%v\" type but was initialized to a different type \"%v\"", needed, trr.Type)
```

Reply:
sure! Thanks!